File: scripts/fetch-instagram.ts
import { getDb, ensureMigrated } from "../src/lib/db";

const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36";
const X_IG_APP_ID = "936619743392459";
const LSD_TOKEN = "AVqbxe3J_YA";

interface InstagramPost {
  shortcode: string;
  type: "Reel" | "Post" | "Carrusel";
  caption: string;
  published_at: string;
  views: number | null;
  likes: number;
  comments: number;
  url: string;
}

function mapTypename(typename: string, productType?: string): "Reel" | "Post" | "Carrusel" {
  if (typename === "GraphSidecar") return "Carrusel";
  if (typename === "GraphVideo" || productType === "clips") return "Reel";
  return "Post";
}

async function fetchUserPosts(username: string): Promise<InstagramPost[]> {
  const profileUrl = new URL("https://www.instagram.com/api/graphql");
  profileUrl.searchParams.set(
    "variables",
    JSON.stringify({ username, render_surface: "PROFILE" })
  );
  profileUrl.searchParams.set("doc_id", "7842794569160958");
  profileUrl.searchParams.set("lsd", LSD_TOKEN);

  const profileRes = await fetch(profileUrl, {
    method: "POST",
    headers: {
      "User-Agent": USER_AGENT,
      "Content-Type": "application/x-www-form-urlencoded",
      "X-IG-App-ID": X_IG_APP_ID,
      "X-FB-LSD": LSD_TOKEN,
      "X-ASBD-ID": "129477",
      "Sec-Fetch-Site": "same-origin",
    },
  });

  if (!profileRes.ok) {
    throw new Error(`Profile fetch failed: ${profileRes.status}`);
  }

  const profileData = await profileRes.json();
  const user = profileData?.data?.user;

  if (!user) {
    throw new Error(`User "${username}" not found`);
  }

  const edges =
    user?.edge_owner_to_timeline_media?.edges ||
    user?.edge_felix_video_timeline?.edges ||
    [];

  const posts: InstagramPost[] = [];

  for (const edge of edges) {
    const node = edge.node;
    const shortcode = node.shortcode;

    try {
      const detail = await fetchPostDetail(shortcode);
      posts.push(detail);
      await new Promise((r) => setTimeout(r, 500));
    } catch (err) {
      console.warn(`  Skipping ${shortcode}: ${(err as Error).message}`);
    }
  }

  return posts;
}

async function fetchPostDetail(shortcode: string): Promise<InstagramPost> {
  const graphql = new URL("https://www.instagram.com/api/graphql");
  graphql.searchParams.set("variables", JSON.stringify({ shortcode }));
  graphql.searchParams.set("doc_id", "10015901848480474");
  graphql.searchParams.set("lsd", LSD_TOKEN);

  const response = await fetch(graphql, {
    method: "POST",
    headers: {
      "User-Agent": USER_AGENT,
      "Content-Type": "application/x-www-form-urlencoded",
      "X-IG-App-ID": X_IG_APP_ID,
      "X-FB-LSD": LSD_TOKEN,
      "X-ASBD-ID": "129477",
      "Sec-Fetch-Site": "same-origin",
    },
  });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }

  const json = await response.json();
  const item = json?.data?.xdt_shortcode_media;

  if (!item) {
    throw new Error("No data returned");
  }

  const typename = item.__typename as string;
  const productType = item.product_type as string | undefined;
  const type = mapTypename(typename, productType);

  const caption =
    item.edge_media_to_caption?.edges?.[0]?.node?.text?.slice(0, 150) || "";

  const views = item.video_play_count || item.video_view_count || null;
  const likes = item.edge_media_preview_like?.count || 0;
  const comments = item.edge_media_to_parent_comment?.count || 0;

  const timestamp = item.taken_at_timestamp;
  const published_at = new Date(timestamp * 1000).toISOString();

  return {
    shortcode,
    type,
    caption,
    published_at,
    views,
    likes,
    comments,
    url: `https://www.instagram.com/p/${shortcode}/`,
  };
}

async function main() {
  const username = process.argv[2];
  if (!username) {
    console.error("Usage: npx tsx scripts/fetch-instagram.ts <username>");
    process.exit(1);
  }

  console.log(`Fetching Instagram posts for @${username}...`);

  const posts = await fetchUserPosts(username);
  console.log(`Found ${posts.length} posts`);

  const sql = getDb();
  await ensureMigrated();

  for (const post of posts) {
    const totalInteractions = post.likes + post.comments;
    const engagementRate = post.views
      ? (totalInteractions / post.views) * 100
      : null;

    await sql`
      INSERT INTO instagram_posts
        (url, shortcode, type, caption, published_at, views, likes, comments, engagement_rate, fetched_at)
      VALUES
        (${post.url}, ${post.shortcode}, ${post.type}, ${post.caption},
         ${post.published_at}, ${post.views}, ${post.likes}, ${post.comments},
         ${engagementRate}, NOW())
      ON CONFLICT(url) DO UPDATE SET
        views = EXCLUDED.views,
        likes = EXCLUDED.likes,
        comments = EXCLUDED.comments,
        engagement_rate = EXCLUDED.engagement_rate,
        fetched_at = NOW()
    `;
  }

  console.log(`Saved ${posts.length} posts to database`);
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: scripts/fetch-instagram.py
#!/usr/bin/env python3
"""
Fetch ALL Instagram posts for a profile and save to Neon (Postgres).
Uses instaloader session cookies + Instagram v1 feed API with pagination.

Requires DATABASE_URL_UNPOOLED (or DATABASE_URL) in env. sync-all.sh sources
.env.local before running this script. Schema must already exist (created by
src/lib/db.ts:ensureMigrated() — first hit of any /api route migrates it).
"""

import instaloader
import json
import sys
import time
import os
from datetime import datetime

import psycopg2
import psycopg2.extras

# Prefer unpooled connection for one-shot scripts (avoids pooler session limits).
# Fall back to the pooled URL if only that is set.
DATABASE_URL = os.environ.get("DATABASE_URL_UNPOOLED") or os.environ.get("DATABASE_URL")

HEADERS = {
    "User-Agent": "Instagram 275.0.0.27.98 Android",
    "X-IG-App-ID": "936619743392459",
}

def init_db():
    if not DATABASE_URL:
        raise SystemExit(
            "❌ DATABASE_URL_UNPOOLED / DATABASE_URL not set. "
            "Run via sync-all.sh (which sources .env.local) or export it manually."
        )
    # Schema is owned by src/lib/db.ts:ensureMigrated(). We only OPEN the connection
    # here; we don't try to CREATE TABLE because the column types differ from SQLite
    # (SERIAL vs AUTOINCREMENT, TIMESTAMPTZ vs DATETIME, DOUBLE PRECISION vs REAL).
    return psycopg2.connect(DATABASE_URL)

def get_session():
    L = instaloader.Instaloader(
        download_pictures=False, download_videos=False,
        download_video_thumbnails=False, save_metadata=False,
    )
    L.load_session_from_file('albertodviejo')
    return L.context._session

def map_media_type(media_type, product_type=None):
    # media_type: 1=photo, 2=video, 8=carousel
    if media_type == 8:
        return "Carrusel"
    if media_type == 2:
        return "Reel"
    return "Post"

def get_user_id(session, username):
    resp = session.get(
        "https://i.instagram.com/api/v1/users/web_profile_info/",
        params={"username": username},
        headers=HEADERS,
    )
    if resp.status_code != 200:
        raise Exception(f"Profile fetch failed: {resp.status_code}")
    data = resp.json()
    user = data["data"]["user"]
    user_id = user["id"]
    followers = user.get("edge_followed_by", {}).get("count", 0)
    total_posts = user.get("edge_owner_to_timeline_media", {}).get("count", 0)
    print(f"📊 @{username}: {followers} followers, {total_posts} posts")
    return user_id

def fetch_all_posts(session, user_id):
    posts = []
    max_id = ""
    page = 1

    while True:
        params = {"count": 33}
        if max_id:
            params["max_id"] = max_id

        resp = session.get(
            f"https://i.instagram.com/api/v1/feed/user/{user_id}/",
            params=params,
            headers=HEADERS,
        )

        if resp.status_code != 200:
            print(f"  ❌ Feed error: {resp.status_code}")
            break

        data = resp.json()
        items = data.get("items", [])

        if not items:
            break

        for item in items:
            shortcode = item.get("code", "")
            media_type = item.get("media_type", 1)
            product_type = item.get("product_type")
            post_type = map_media_type(media_type, product_type)

            caption_obj = item.get("caption")
            caption = (caption_obj.get("text", "")[:150] if caption_obj else "").replace('\n', ' ')

            timestamp = item.get("taken_at", 0)
            published_at = datetime.utcfromtimestamp(timestamp).isoformat() if timestamp else ""

            views = item.get("play_count") or item.get("view_count")
            likes = item.get("like_count", 0)
            comments = item.get("comment_count", 0)

            # Thumbnail
            candidates = item.get("image_versions2", {}).get("candidates", [])
            thumbnail_url = candidates[0]["url"] if candidates else None

            total = likes + comments
            eng_rate = round((total / views * 100), 2) if views and views > 0 else None

            posts.append({
                "url": f"https://www.instagram.com/p/{shortcode}/",
                "shortcode": shortcode,
                "type": post_type,
                "caption": caption,
                "published_at": published_at,
                "views": views,
                "likes": likes,
                "comments": comments,
                "engagement_rate": eng_rate,
                "thumbnail_url": thumbnail_url,
            })

        print(f"  📄 Page {page}: {len(items)} posts (total: {len(posts)})")

        more = data.get("more_available", False)
        next_max_id = data.get("next_max_id")

        if not more or not next_max_id:
            break

        max_id = next_max_id
        page += 1
        time.sleep(2)

    return posts

def save_to_db(conn, posts):
    """Upsert posts to Neon. Postgres ON CONFLICT mirrors the previous SQLite logic."""
    saved = 0
    with conn.cursor() as cursor:
        for p in posts:
            cursor.execute(
                """
                INSERT INTO instagram_posts (
                    url, shortcode, type, caption, published_at,
                    views, likes, comments, engagement_rate, thumbnail_url, fetched_at
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, NOW())
                ON CONFLICT (url) DO UPDATE SET
                    views = EXCLUDED.views,
                    likes = EXCLUDED.likes,
                    comments = EXCLUDED.comments,
                    engagement_rate = EXCLUDED.engagement_rate,
                    thumbnail_url = EXCLUDED.thumbnail_url,
                    fetched_at = NOW()
                """,
                (
                    p["url"], p["shortcode"], p["type"], p["caption"],
                    p["published_at"], p["views"], p["likes"], p["comments"],
                    p["engagement_rate"], p["thumbnail_url"],
                ),
            )
            saved += 1
    conn.commit()
    return saved

if __name__ == "__main__":
    username = sys.argv[1].replace('@', '') if len(sys.argv) > 1 else "albertodviejo"

    print(f"🔄 Fetching ALL Instagram posts for @{username}...")

    session = get_session()
    print("✅ Session loaded")

    user_id = get_user_id(session, username)
    posts = fetch_all_posts(session, user_id)

    print(f"\n📥 Found {len(posts)} posts total")

    if posts:
        conn = init_db()
        saved = save_to_db(conn, posts)
        conn.close()
        print(f"💾 Saved {saved} posts to DB")

        reels = sum(1 for p in posts if p["type"] == "Reel")
        carruseles = sum(1 for p in posts if p["type"] == "Carrusel")
        statics = sum(1 for p in posts if p["type"] == "Post")
        print(f"📊 {reels} reels · {carruseles} carruseles · {statics} posts")
    else:
        print("⚠️ No posts found")
